src/test/java/com/browserstack/examples/providers/WebDriverDataProvider.java

```diff
@@ -0,0 +1,30 @@
+package com.browserstack.examples.providers;
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.testng.annotations.DataProvider;
+
+import com.browserstack.examples.config.Platform;
+import com.browserstack.examples.config.WebDriverFactory;
+
+/**
+ * Created with IntelliJ IDEA.
+ *
+ * @author Anirudha Khanna
+ */
+public class WebDriverDataProvider {
+
+    @DataProvider(name = "webDriverProvider", parallel = true)
```

Comment:
Main reason to use DataProvider for providing the WebDriver is also to tap into the parallelization support that TestNG provides to DataProvider.

This is a matter of opinion where I feel TestNG and JUnit4 force you to think of parameters passed to a function as only Data. This then forces one to think about the WebDriver test environments as Data. The links that you share map test parameters to only Data but in the case of Cross Browser testing Test parameters also need to be thought of as the platforms where the test should be run, (aka webdriverio and nightwatchjs).

In terms of using multiple DataProviders for Data driven testing, the links that you share seem to make it very easy to add extend the DataProvider method to add the additional data to it as well since the tests will need to be run in the m x n combinations where m = number of platforms and n = different data params to be tested.